### PR TITLE
[runtimes] Probe for -nostdlib++ and -nostdinc++ with the C compiler

### LINKED ISSUE #90332: [runtimes] [openmp] `-nostdlib++` added to `CMAKE_REQUIRED_FLAGS` breaks hwloc detection for OMPT
To reproduce:

```
$ cmake -G Ninja -DLLVM_ENABLE_RUNTIMES=openmp -DLIBOMP_USE_HWLOC=ON ${path_to_llvm_project}/runtimes
[…]
-- Looking for hwloc.h
-- Looking for hwloc.h - not found
-- Looking for hwloc_topology_init in /usr/lib64/libhwloc.so
-- Looking for hwloc_topology_init in /usr/lib64/libhwloc.so - not found
-- LIBOMP: Could not find hwloc
CMake Error at /home/mgorny/git/llvm-project/openmp/runtime/cmake/LibompUtils.cmake:26 (message):
  LIBOMP: Hwloc requested but not available
Call Stack (most recent call first):
  /home/mgorny/git/llvm-project/openmp/runtime/CMakeLists.txt:387 (libomp_error_say)


-- Configuring incomplete, errors occurred!
```

`CMakeFiles/CMakeConfigureLog.yaml` hints at the problem:

```yaml
  -
    kind: "try_compile-v1"
    backtrace:
      - "/usr/share/cmake/Modules/CheckIncludeFile.cmake:90 (try_compile)"
      - "/home/mgorny/git/llvm-project/openmp/runtime/cmake/config-ix.cmake:351 (check_include_file)"
      - "/home/mgorny/git/llvm-project/openmp/runtime/CMakeLists.txt:276 (include)"
    checks:
      - "Looking for hwloc.h"
    directories:
      source: "/home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY"
      binary: "/home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY"
    cmakeVariables:
      CMAKE_C_FLAGS: " -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections"
      CMAKE_C_FLAGS_DEBUG: "-g"
      CMAKE_EXE_LINKER_FLAGS: ""
      CMAKE_MODULE_PATH: "/home/mgorny/git/llvm-project/openmp/runtime/cmake;/home/mgorny/git/llvm-project/openmp/cmake;/home/mgorny/git/llvm-project/openmp/../cmake/Modules;/home/mgorny/git/llvm-project/runtimes/cmake;/home/mgorny/git/llvm-project/runtimes/cmake/modules;/home/mgorny/git/llvm-project/runtimes/../cmake;/home/mgorny/git/llvm-project/runtimes/../cmake/Modules;/home/mgorny/git/llvm-project/runtimes/../llvm/cmake;/home/mgorny/git/llvm-project/runtimes/../llvm/cmake/modules"
    buildResult:
      variable: "LIBOMP_HAVE_HWLOC_H"
      cached: true
      stdout: |
        Change Dir: '/home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY'
        
        Run Build Command(s): /usr/bin/ninja -v cmTC_5eb96
        [1/2] /usr/bin/cc   -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections  -nostdlib++ -nostdinc++ -o CMakeFiles/cmTC_5eb96.dir/CheckIncludeFile.c.o -c /home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY/CheckIncludeFile.c
        FAILED: CMakeFiles/cmTC_5eb96.dir/CheckIncludeFile.c.o 
        /usr/bin/cc   -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -Wno-comment -fdiagnostics-color -Wall -Wcast-qual -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -Wno-maybe-uninitialized -fno-semantic-interposition -fdata-sections  -nostdlib++ -nostdinc++ -o CMakeFiles/cmTC_5eb96.dir/CheckIncludeFile.c.o -c /home/mgorny/git/llvm-project/build/CMakeFiles/CMakeScratch/TryCompile-EEYwvY/CheckIncludeFile.c
        cc: error: unrecognized command-line option ‘-nostdlib++’
        ninja: build stopped: subcommand failed.
        
      exitCode: 1
```

i.e. the `nostdlib++` flags gets appended to *both* C and C++ compiler flags, effectively breaking all compilations done in C mode.

CC @mstorsjo as the one who added the logic

### PR DESCRIPTION
**Pull request by mstorsjo**
While these flags semantically are relevant only for C++, we do add them to CMAKE_REQUIRED_FLAGS if they are detected. All flags in that variable are used both when testing compilation of C and C++ (and for detecting libraries, which uses the C compiler driver).

Therefore, to be sure we safely can add the flags to CMAKE_REQUIRED_FLAGS, test for the option with the C language.

This should fix compilation with GCC; newer versions of GCC do support the -nostdlib++ option, but it's only supported by the C++ compiler driver, not the C driver. (However, many builds of GCC also do accept the option with the C driver, if GCC was compiled with Ada support enabled, see [1]. That's why this issue isn't noticed in all configurations with GCC.)

Clang does support these options in both C and C++ driver modes.

This should fix #90332.

[1] https://github.com/llvm/llvm-project/issues/90332#issuecomment-2325099254

### CODE REVIEW

File: libcxx/cmake/config-ix.cmake [hunk 1]
```diff
@@ -38,9 +38,13 @@ check_cxx_compiler_flag(-nolibc CXX_SUPPORTS_NOLIBC_FLAG)
...
 # required for the link to go through. We remove sanitizers from the
 # configuration checks to avoid spurious link errors.
+#
+# Adding flags to CMAKE_REQUIRED_FLAGS will include the flags both when testing
+# compilation of C and C++. Therefore test to make sure that the flags are
+# supported by the C compiler driver, before deciding to include them.
 
-check_cxx_compiler_flag(-nostdlib++ CXX_SUPPORTS_NOSTDLIBXX_FLAG)
-if (CXX_SUPPORTS_NOSTDLIBXX_FLAG)
+check_c_compiler_flag(-nostdlib++ C_SUPPORTS_NOSTDLIBXX_FLAG)
```
Comment (ldionne) [inline libcxx/cmake/config-ix.cmake:46]: Do we compile any C sources in libc++? Why do we care what flags the C compiler supports for libc++ and libc++abi? (I understand why it matters for libunwind where we do have C sources).

The reason I'm asking is that there are some simplifications we can make if we assume that all supported compilers accept `-nostdlib++`, and this patch makes it impossible to make that assumption.

Comment (mstorsjo) [reply libcxx/cmake/config-ix.cmake:46]: The problem is that even if we don’t use any C sources outright, cmake uses C compilation tests when you do a check for whether a library exists. And these are subtly broken if building with GCC right now. (To make matters worse, not all GCC versions are affected. But that’s a bug - GCC intends to reject these options in the C frontend.) IIRC one of the issues mentioned in the linked bug was about library detection in libcxxabi.

Comment (ldionne) [reply libcxx/cmake/config-ix.cmake:46]: > The problem is that even if we don’t use any C sources outright, cmake uses C compilation tests when you do a check for whether a library exists.

Do you mean these checks? https://github.com/llvm/llvm-project/blob/aa02b76b1a4c2df1a701fbd8a8378d1cd946e70a/libcxx/cmake/config-ix.cmake#L130

Comment (mstorsjo) [reply libcxx/cmake/config-ix.cmake:46]: Yes, exactly those; any of those basic compiler tests, `check_library_exists` (used a couple of times in libcxxabi and libcxx), `check_include_file` (not used in libcxxabi/libcxx, but used in openmp, and also struck by the same issue, in #90332), are executed as plain C compilation, and they will fail if we have stuck C++-only flags in `CMAKE_REQUIRED_FLAGS`.

Comment (mstorsjo) [reply libcxx/cmake/config-ix.cmake:46]: Ping

**Review by ldionne — APPROVED** [review]

**Comment by mstorsjo** [comment]
FWIW, I'd like to backport this to 19.x if possible. But I'll wait for a little while to see whether this turns up any issues, before filing the backport request.

**Comment by vitalybuka** [comment]
Actually this PR is the case of https://lab.llvm.org/buildbot/#/builders/164/builds/3908, not https://github.com/llvm/llvm-project/pull/113491

**Comment by mstorsjo** [comment]
> Actually this PR is the case of https://lab.llvm.org/buildbot/#/builders/164/builds/3908, not #113491

Sorry about that.

Are you able to dig up any of the cmake configure logs from these builds, so that we can figure out how this ends up breaking things? Because as in most of these cases, the visible output in the buildbot logs mostly say that some configure checks end up going the wrong way, but we can't really see why.

**Comment by mstorsjo** [comment]
> > Actually this PR is the case of https://lab.llvm.org/buildbot/#/builders/164/builds/3908, not #113491
> 
> Sorry about that.
> 
> Are you able to dig up any of the cmake configure logs from these builds, so that we can figure out how this ends up breaking things? Because as in most of these cases, the visible output in the buildbot logs mostly say that some configure checks end up going the wrong way, but we can't really see why.

@vitalybuka Are you able to pull out some details to help figure out what this really broke here?

**Comment by mstorsjo** [comment]
> > > Actually this PR is the case of https://lab.llvm.org/buildbot/#/builders/164/builds/3908, not #113491
> > 
> > 
> > Sorry about that.
> > Are you able to dig up any of the cmake configure logs from these builds, so that we can figure out how this ends up breaking things? Because as in most of these cases, the visible output in the buildbot logs mostly say that some configure checks end up going the wrong way, but we can't really see why.
> 
> @vitalybuka Are you able to pull out some details to help figure out what this really broke here?

Ping @vitalybuka - can you help out with retrying this PR and providing some details about what went wrong, so that we can have a path towards merging this again?

**Comment by h-vetinari** [comment]
Ping @vitalybuka

**Comment by vitalybuka** [comment]
> > > > Actually this PR is the case of https://lab.llvm.org/buildbot/#/builders/164/builds/3908, not #113491
> > > 
> > > 
> > > Sorry about that.
> > > Are you able to dig up any of the cmake configure logs from these builds, so that we can figure out how this ends up breaking things? Because as in most of these cases, the visible output in the buildbot logs mostly say that some configure checks end up going the wrong way, but we can't really see why.
> > 
> > 
> > @vitalybuka Are you able to pull out some details to help figure out what this really broke here?
> 
> Ping @vitalybuka - can you help out with retrying this PR and providing some details about what went wrong, so that we can have a path towards merging this again?

Sure, I can retry the build locally, but would be more convenient for you to just reproduce according to  https://github.com/google/sanitizers/wiki/SanitizerBotReproduceBuild ?
libcxx step is very early, so no need to wait for entire build.

I guess you can just copy cmake invocation from 
https://lab.llvm.org/buildbot/#/builders/sanitizer-x86_64-linux-bootstrap-msan/
you need steps:
1.  stage1 build all
2.  stage2/msan build libcxx

**Comment by vitalybuka** [comment]
The problem is that some of check_cxx_source_compiles invocation does:

```
cat t.cpp 
int main() {
 }

/usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/bin/clang++ -fsanitize=memory  -nostdlib++ -nostdinc++  -fuse-ld=lld t.cpp 

>> referenced by ubsan_type_hash_itanium.cpp:245 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:245)
>>>               ubsan_type_hash_itanium.cpp.o:(__ubsan::checkDynamicType(void*, void*, unsigned long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a

ld.lld: error: undefined symbol: typeinfo for __cxxabiv1::__class_type_info
>>> referenced by ubsan_type_hash_itanium.cpp:245 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:245)
>>>               ubsan_type_hash_itanium.cpp.o:(__ubsan::checkDynamicType(void*, void*, unsigned long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
>>> referenced by ubsan_type_hash_itanium.cpp:0 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:0)
>>>               ubsan_type_hash_itanium.cpp.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
>>> referenced by ubsan_type_hash_itanium.cpp:141 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:141)
>>>               ubsan_type_hash_itanium.cpp.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
>>> referenced 2 more times

and so on
```

So -nostdlib++ removes libc++ but does not remove msan_cxx.a.

My guess we need to fix the driver to do so.

**Comment by vitalybuka** [comment]
```
diff --git a/clang/lib/Driver/SanitizerArgs.cpp b/clang/lib/Driver/SanitizerArgs.cpp
index 89f1215afd0c..234c28bfd2c3 100644
--- a/clang/lib/Driver/SanitizerArgs.cpp
+++ b/clang/lib/Driver/SanitizerArgs.cpp
@@ -1083,10 +1083,11 @@ SanitizerArgs::SanitizerArgs(const ToolChain &TC,
                    options::OPT_fno_sanitize_link_runtime, LinkRuntimes);
 
   // Parse -link-cxx-sanitizer flag.
-  LinkCXXRuntimes = Args.hasArg(options::OPT_fsanitize_link_cxx_runtime,
-                                options::OPT_fno_sanitize_link_cxx_runtime,
-                                LinkCXXRuntimes) ||
-                    D.CCCIsCXX();
+  LinkCXXRuntimes =
+      Args.hasFlag(options::OPT_fsanitize_link_cxx_runtime,
+                  options::OPT_fno_sanitize_link_cxx_runtime,
+                  LinkCXXRuntimes) ||
+      (D.CCCIsCXX() && !Args.hasArg(clang::driver::options::OPT_nostdlibxx))  ;
 
   NeedsMemProfRt = Args.hasFlag(options::OPT_fmemory_profile,
                                 options::OPT_fmemory_profile_EQ,
```

@petrhosek  `Args.hasArg` looks like a typo in https://reviews.llvm.org/D65029 ?

**Comment by mstorsjo** [comment]
> The problem is that some of check_cxx_source_compiles invocation does:
> 
> ```
> cat t.cpp 
> int main() {
>  }
> 
> /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/bin/clang++ -fsanitize=memory  -nostdlib++ -nostdinc++  -fuse-ld=lld t.cpp 
> 
> >> referenced by ubsan_type_hash_itanium.cpp:245 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:245)
> >>>               ubsan_type_hash_itanium.cpp.o:(__ubsan::checkDynamicType(void*, void*, unsigned long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
> 
> ld.lld: error: undefined symbol: typeinfo for __cxxabiv1::__class_type_info
> >>> referenced by ubsan_type_hash_itanium.cpp:245 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:245)
> >>>               ubsan_type_hash_itanium.cpp.o:(__ubsan::checkDynamicType(void*, void*, unsigned long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
> >>> referenced by ubsan_type_hash_itanium.cpp:0 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:0)
> >>>               ubsan_type_hash_itanium.cpp.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
> >>> referenced by ubsan_type_hash_itanium.cpp:141 (/usr/local/google/home/vitalybuka/tmp/bot/llvm-project/compiler-rt/lib/ubsan/ubsan_type_hash_itanium.cpp:141)
> >>>               ubsan_type_hash_itanium.cpp.o:(isDerivedFromAtOffset(__cxxabiv1::__class_type_info const*, __cxxabiv1::__class_type_info const*, long)) in archive /usr/local/google/home/vitalybuka/tmp/bot/llvm_build0/lib/clang/20/lib/x86_64-unknown-linux-gnu/libclang_rt.msan_cxx.a
> >>> referenced 2 more times
> 
> and so on
> ```
> 
> So -nostdlib++ removes libc++ but does not remove msan_cxx.a.
> 
> My guess we need to fix the driver to do so.

Thanks! So this means, with this patch, we conclude that `-nostdlib++` doesn't work, and this causes the build to fail later for other reasons? (The buildbot logs seems to be expired at this point.)

But I find this a bit confusing still, because this patch should make the probing to be done with `$CC`, not `$CXX`, so then the test case in your example should be using `clang`, not `clang++`? Or is there some longer series of changed conclusions here?

**Comment by vitalybuka** [comment]
> But I find this a bit confusing still, because this patch should make the probing to be done with `$CC`, not `$CXX`, so then the test case in your example should be using `clang`, not `clang++`? Or is there some longer series of changed conclusions here?

Error is from this check: https://github.com/llvm/llvm-project/blob/1235a93fae60bed5814e918dd8608097d9302a59/llvm/cmake/modules/HandleLLVMOptions.cmake#L410

**Comment by mstorsjo** [comment]
Ok, now I tried this myself (and indeed, it wasn't very hard to set up - sorry for not trying it earlier).

Here's the effect on the configure run, caused by this patch:
```diff
--- log-good-truncated  2024-12-18 13:59:16.334711800 +0200
+++ log-bad     2024-12-18 13:58:07.287917504 +0200
@@ -29,10 +29,19 @@
 -- Could NOT find Clang (missing: Clang_DIR)
 -- Performing Test CXX_SUPPORTS_UNWINDLIB_EQ_NONE_FLAG
 -- Performing Test CXX_SUPPORTS_UNWINDLIB_EQ_NONE_FLAG - Failed
--- Performing Test CXX_SUPPORTS_NOSTDLIBXX_FLAG
--- Performing Test CXX_SUPPORTS_NOSTDLIBXX_FLAG - Failed
--- Performing Test CXX_SUPPORTS_NOSTDINCXX_FLAG
--- Performing Test CXX_SUPPORTS_NOSTDINCXX_FLAG - Success
+-- Performing Test C_SUPPORTS_NOSTDLIBXX_FLAG
+-- Performing Test C_SUPPORTS_NOSTDLIBXX_FLAG - Success
+-- Performing Test C_SUPPORTS_NOSTDINCXX_FLAG
+-- Performing Test C_SUPPORTS_NOSTDINCXX_FLAG - Success
 -- Linker detection: LLD
 -- Performing Test CXX_SUPPORTS_CUSTOM_LINKER
--- Performing Test CXX_SUPPORTS_CUSTOM_LINKER - Success
+-- Performing Test CXX_SUPPORTS_CUSTOM_LINKER - Failed
+CMake Error at /home/martin/code/llvm-project/llvm/cmake/modules/HandleLLVMOpti
ons.cmake:412 (message):
+  Host compiler does not support '-fuse-ld=lld'.  Please make sure that 'lld'
+  is installed and that your host compiler can compile a simple program when
+  given the option '-fuse-ld=lld'.
+Call Stack (most recent call first):
+  CMakeLists.txt:179 (include)
+
+
+-- Configuring incomplete, errors occurred!
```

So previously, `CXX_SUPPORTS_NOSTDLIBXX_FLAG` failed, so we didn't add `-nostdlib++`. It failed due to the issue that does get fixed by https://github.com/llvm/llvm-project/pull/120370. By applying this change, we happen to go around that issue, so we end up adding `-nostdlib++`, even when it fails to link in C++ mode.

So, indeed, https://github.com/llvm/llvm-project/pull/120370 seems like a proper fix for this. Thanks!

(Ideally, we probably should probe for support for these flags separately for C and C++ mode, but it would require CMake to have something like `CMAKE_REQUIRED_FLAGS` separately for C and C++, which it doesn't - that's the whole root cause for this mess.)

**Comment by mstorsjo** [comment]
Ok, so with https://github.com/llvm/llvm-project/pull/120370 landed, I guess we could consider trying to reland this.

By doing that, we would however break doing msan+libcxx builds unless using the very latest Clang (while libcxx in general supports building with the last two stable releases of Clang). Do you think this is an issue, or are we ok with implicitly requiring the very latest for this particular configuration?

**Comment by h-vetinari** [comment]
We could backport #120370 so that at least clang 19.1.x still becomes compatible again (post-reland)?

**Comment by vitalybuka** [comment]
#120370 was reverted, because libc++ tests expose major problem:
Tests and likely many users can have `-nostdlib++ -lc++`.
The if we remove c++ sanitizer runtimes from linking it will break them and force to add `-fsanitize-link-c++-runtime`

I will reland unrelated parts of #120370, and think for other solutions.

**Comment by mstorsjo** [comment]
> #120370 was reverted, because libc++ tests expose major problem:
> Tests and likely many users can have `-nostdlib++ -lc++`.
> The if we remove c++ sanitizer runtimes from linking it will break them and force to add `-fsanitize-link-c++-runtime`

That's indeed a problem... However the libc++ test failure, as far as I can see, seems to be that there was a libc++ test that used to fail when built with sanitizers, but that test now no longer fail in that configuration. (I'm not entirely sure why it did use to fail though - but I guess that discussion belongs in #120370 - I'll comment more on it there.)

**Comment by vitalybuka** [comment]
> > #120370 was reverted, because libc++ tests expose major problem:
> > Tests and likely many users can have `-nostdlib++ -lc++`.
> > The if we remove c++ sanitizer runtimes from linking it will break them and force to add `-fsanitize-link-c++-runtime`
> 
> That's indeed a problem... However the libc++ test failure, as far as I can see, seems to be that there was a libc++ test that used to fail when built with sanitizers, but that test now no longer fail in that configuration. (I'm not entirely sure why it did use to fail though - but I guess that discussion belongs in #120370 - I'll comment more on it there.)

That's apply to asan,msan

`-nostdlib++ -lc++` is relevant for many ubsan tests https://lab.llvm.org/buildbot/#/builders/25/builds/5017

**Comment by mstorsjo** [comment]
> > > #120370 was reverted, because libc++ tests expose major problem:
> > > Tests and likely many users can have `-nostdlib++ -lc++`.
> > > The if we remove c++ sanitizer runtimes from linking it will break them and force to add `-fsanitize-link-c++-runtime`
> > 
> > 
> > That's indeed a problem... However the libc++ test failure, as far as I can see, seems to be that there was a libc++ test that used to fail when built with sanitizers, but that test now no longer fail in that configuration. (I'm not entirely sure why it did use to fail though - but I guess that discussion belongs in #120370 - I'll comment more on it there.)
> 
> That's apply to asan,msan
> 
> `-nostdlib++ -lc++` is relevant for many ubsan tests https://lab.llvm.org/buildbot/#/builders/25/builds/5017

Ah, I see, yeah, there's clearly not just one single test failing there. I mirrored part of this conversation into #120370 where it may be more on topic regarding the fallout of that change.

**Pull request merged**